### PR TITLE
Remove nodeIDs by user action or after timeout

### PR DESCRIPTION
**Pull request by ricmail85**
It is possible now to remove a nodeID (for instance if it is dead) by a specific user action or after a timeout. 
I also implemented a while loop in the mesh.write routine in order to try several time to send a message and stop if success. 
This because there are several open issues on master-node communications (especially master to node). It happens that the master send a message (correctly received by the node or not) but it results always false to the master.

### CODE REVIEW

**Comment by Avamander** [comment]
I think the remove function should be excluded by the C preprocessor with if(n)def when it's not on the master. In addition to that I think both the dead node checking/pinging and message retry should be also done only when a compile time flag is set. Doing that allows the user to make sure no unnecessary functions are compiled/uploaded and keeps identical compatibility with older versions (as the new feature has to be enabled). (There's also an issue with bad optimization of unused functions open in the Arduino issue tracker but not sure that's relevant)

**Comment by ricmail85** [comment]
Hi, I agree with you but there is no define for the master node. I mean, how can I check if the code is running on the master or not before definition?

**Comment by Avamander** [comment]
ifndef nodeID?

**Comment by ricmail85** [comment]
where is it defined?

**Comment by Avamander** [comment]
It is usually defined in the non-master sketches, see the examples.

**Comment by ricmail85** [comment]
Thanks for the reply. So if I define nodeID in the arduino sketch, the definition is available also in  the RF24Mesh.cpp library? Sorry but I'm not a C++ expert.
bye

**Comment by ricmail85** [comment]
Is it the same as MESH_NOMASTER?

**Comment by Avamander** [comment]
MESH_NOMASTER doesn't remove anything that might be useful on non-master nodes?

About the nodeID definition, it is available when it's defined before the file is included I think. But there are cases of "nodeID" in RF24Mesh that can not be redefined like that, one possibility is renaming the usages that must not be rewritten, another would be creating a another new define flag.

**Comment by ricmail85** [comment]
Hi, after testing more I think there is no need of the ifndef NodeID check to avoid writing unnecessary code to the node. By looking at the memory usage, the size is exactly the same with the new routines commented out or not. And they can be executed only on the master since they check for the mesh_address=0 and operate only if mesh_address=0.

If you agree we can marge as it is since I've already tested both on master and nodes.

**Comment by Avamander** [comment]
RF24Mesh_example.ino New:
```
Sketch uses 10534 bytes (32%) of program storage space. Maximum is 32256 bytes.
Global variables use 695 bytes (33%) of dynamic memory, leaving 1353 bytes for local variables. Maximum is 2048 bytes.
```
Old:
```
Sketch uses 10312 bytes (31%) of program storage space. Maximum is 32256 bytes.
Global variables use 695 bytes (33%) of dynamic memory, leaving 1353 bytes for local variables. Maximum is 2048 bytes.
```

There definitely is a difference and I think the difference should only exist when the user defines that it's needed.

**Comment by ricmail85** [comment]
Ok in the new version if the user defined the MESH_NOMASTER option, the new routines are not compiled and uploaded, saving memory. Since these routines must be run only on master, I think the best way to avoid unnecessary routines is to use the MESH_NOMASTER definition.

File: RF24Mesh.cpp [hunk 1]
```diff
@@ -88,13 +86,91 @@ uint8_t RF24Mesh::update(){
...
+#endif
+
 bool RF24Mesh::write(uint16_t to_node, const void* data, uint8_t msg_type, size_t size ){
     if(mesh_address == MESH_DEFAULT_ADDRESS){ return 0; }
+	
+	bool _success=false;
 	RF24NetworkHeader header(to_node,msg_type);	
-	return network.write(header,data,size);	
+	//return network.write(header,data,size);	
+	uint32_t ltime = millis();
```
Comment (Avamander) [inline RF24Mesh.cpp]: Though this part still gets compiled and included. Maybe "#ifdef MESH_WRITE_RETRY"? It also makes this function blocking longer until sent.

**Comment by Avamander** [comment]
Thanks a lot for your commit, it's a nice improvement. Though see my comment in the code. The write function should also stay intact unless defined otherwise because the additional delay might not be wanted.

Overall I think if you wish you could also make a new example that uses the new define.

**Comment by ricmail85** [comment]
Right. I removed the multiple tentatives on mesh write

**Comment by Avamander** [comment]
Okay, I think now everything you added should be available by defining MESH_WRITE_RETRY. Let me know what you think.

**Comment by ricmail85** [comment]
Ok I agree. So the user can choose to retry or not.

**Comment by Avamander** [comment]
I added it as a overall flag to either use the new features or not.

**Comment by ricmail85** [comment]
ok but MESH_WRITE_RETRY is a different thing with respect to MESH_NOMASTER. Do not mix them.
See my last commit

**Comment by Avamander** [comment]
Now all examples still grow in size because MESH_NOMASTER is not defined in them, the new code should be opt-in not opt-out.

**Comment by ricmail85** [comment]
Yes but also the DHCP function, for instance, is not necessary for nodes but is implemented through the MESH_NOMASTER flag...

**Comment by Avamander** [comment]
So you think the node example should have MESH_NOMASTER defined?

**Comment by ricmail85** [comment]
yes all nodes, in principle, should use the MESH_NOMASTER flag to not waste memory. And I think it's better to let the user decide if include this flag or not. 
Thus, by default, all routines are available to the user sketch.If the user want to save space, the flag MESH_NOMASTER has to be used

**Comment by Avamander** [comment]
But you can't use the flag on the master node thus no meaningful way to exclude the unnecessary code exists.

**Comment by ricmail85** [comment]
For the node the user has to uncomment the MESH_NOMASTER flag in RF24Mesh_config. This is the way TMRh20 implemented such a flag

**Comment by Avamander** [comment]
Here lies the problem. The "solution" is tedious library modification if you wish to compile for master and nodes simultaneously. It will not work. In addition to that, there is no proper way to define that one does not wish to use the new code in their master, that has to be fixed.

**Comment by danhajduk** [comment]
i was just trying to solve this issue. it is working for me , thanks

**Comment by carlosDY** [comment]
Can you include an example of how to use it ??

**Comment by pmaster555** [comment]
good function, why it is not included in the main code?

**Comment by Avamander** [comment]
@pmaster555 

> good function, why it is not included in the main code?

At this point, doesn't match the rest of the code's style and lacks documentation. It also increases resource usage even if the new features are not used. I can't recall if it had any b/w compat issues or not, it's been three years, I think it did.

**Comment by pmaster555** [comment]
function  RF24Mesh::remove(int16_t nodeID)  
causes crashes on esp32 
`CORRUPT HEAP: Bad tail at 0x3ffcbfd0. Expected 0xbaad5678 got 0x001d5607
assertion "head != NULL" failed: file "/home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/heap/multi_heap_poisoning.c", line 245, function: multi_heap_realloc
abort() was called at PC 0x401032a7 on core 1

Backtrace: 0x4008c494:0x3ffb1e30 0x4008c6c5:0x3ffb1e50 0x401032a7:0x3ffb1e70 0x4008c161:0x3ffb1ea0 0x40084b86:0x3ffb1ec0 0x40084bea:0x3ffb1ee0 0x40084f2d:0x3ffb1f00 0x4000bedd:0x3ffb1f20 0x400d7fff:0x3ffb1f40 0x400d81bc:0x3ffb1f60 0x400d1fb8:0x3ffb1f90 0x400d945d:0x3ffb1fb0 0x40088b81:0x3ffb1fd0

Rebooting...
`

**Comment by TMRh20** [comment]
Due to the code and memory requirements of a timeout feature for nodes, this will be left up to the user to implement.

The current code should allow the application to set the address of any inactive nodes to 0 in order to free up the address space, whether by timeout or user action. (see setAddress function)
If the node comes back online, it will just pick up a new address.

**Pull request closed without merging**